cache/clock_cache.cc

```diff
@@ -407,90 +435,76 @@ Status ClockCacheShard::Insert(const Slice& key, uint32_t hash, void* value,
   return s;
 }
 
-Cache::Handle* ClockCacheShard::Lookup(const Slice& key, uint32_t /* hash */) {
+Cache::Handle* ClockCacheShard::Lookup(const Slice& key, uint32_t hash) {
   ClockHandle* h = nullptr;
-  {
-    DMutexLock l(mutex_);
-    h = table_.Lookup(key);
-    if (h != nullptr) {
-      assert(h->IsVisible());
-      if (!h->HasRefs()) {
-        // The entry is in clock since it's in the hash table and has no
-        // external references.
-        ClockRemove(h);
-      }
-      h->Ref();
-      h->SetHit();
-    }
+  h = table_.Lookup(key, hash);
+  if (h != nullptr) {
+    h->SetHit();
```

Comment:
Here it looks like we have three atomic updates where it would be possible to combine into one CAS (more metadata under one atomic field) or maybe two atomic updates (one arithmetic, one bitwise). Something to think about optimizing.

Reply:
We can combine h->SetHit() and the clock update into a single atomic & if we reverse the meaning of the HAS_HIT bit.

The CAS alternative needs an atomic read first, so I'm not sure it'd be better.

Reply:
I'm going to leave your comment as a TODO in the code.